src/jit/flowgraph.cpp

```diff
@@ -4883,6 +4887,26 @@ void Compiler::fgFindJumpTargets(const BYTE* codeAddr, IL_OFFSET codeSize, BYTE*
 
     if (makeInlineObservations)
     {
+        if (retBlocks == 0)
+        {
+            // If there are no return blocks we know it does not return, however if there
+            // return blocks we don't know it returns as it may be counting unreachable code.
+            // So only make the CALLEE_DOES_NOT_RETURN observations if no returns.
+
+            if (!compInlineResult->UsesLegacyPolicy())
+            {
+                // Mark the call node as "no return" as it can impact caller's code quality.
+                impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+
+                if (compInlineResult->IsFailure())
```

Comment:
The `IsFailure` check should come after `NoteBool`, otherwise it is useless.

Considering that `CALLEE_DOES_NOT_RETURN` doesn't currently cause inlining to be aborted right away it might be best to leave this stuff out, the existing "no return" code will do this anyway. All we really need here is to mark the call node as "no return".

Also, it might be good to add an assert in the existing "no return" code to ensure that if `fgFindJumpTargets` marked the call as "no return" then there really are no `BBJ_RETURN` blocks in the method.